### PR TITLE
Waveform equivalence checker.

### PR DESCRIPTION
**Pull request by JakeEdvenson**
Added a rough-draft of the waveform equivalence checker so that we can start testing it out.

### CODE REVIEW

File: scripts/bfasst/compare/waveform.py [hunk 1]
```diff
@@ -0,0 +1,355 @@
...
+import pathlib
+import sys
+import os
+import re
+import fileinput
+from bfasst.compare.base import CompareTool
+from bfasst.status import Status, CompareStatus
+from bfasst.tool import ToolProduct
+from random import randint
+from os.path import exists
```
Comment (jgoeders) [inline scripts/bfasst/compare/waveform.py:12]: Can you use pathlib instead of os.path?

File: scripts/bfasst/compare/waveform.py [hunk 2]
```diff
@@ -0,0 +1,355 @@
...
+    data["oBitsList"].clear()
+    data["totalList"].clear()
+    data["randList"].clear()
+
+
+class Waveform_CompareTool(CompareTool):
+
+    TOOL_WORK_DIR = "waveform"
+    LOG_FILE_NAME = "log.txt"
+    CELLS_SIM = str(pathlib.Path.cwd()) + "/third_party/yosys/techlibs/xilinx/cells_sim.v" + "; "
```
Comment (jgoeders) [inline scripts/bfasst/compare/waveform.py:38]: Use paths.py, not cwd.

File: scripts/bfasst/compare/waveform.py [hunk 3]
```diff
@@ -0,0 +1,355 @@
+import bfasst
```
Comment (jgoeders) [inline scripts/bfasst/compare/waveform.py:1]: Is this file formatted?  Can you format using black?

File: scripts/bfasst/compare/waveform.py [hunk 4]
```diff
@@ -0,0 +1,355 @@
...
+        return(len(data["inputList"]))
+
+    def outputNum(self):
+        return(len(data["outputList"]))
+
+    def totalNum(self):
+        return(len(data["inputList"]) + len(data["outputList"]))
+
+    def fix_file(self, PATH, fileName, isReversed):
+        fin = open(PATH + fileName + ".v", "r")
```
Comment (jgoeders) [inline scripts/bfasst/compare/waveform.py:84]: Use `with` when opening files

File: scripts/bfasst/compare/waveform.py [hunk 5]
```diff
@@ -0,0 +1,355 @@
...
+        fileData = fin.read()
+        if(isReversed):
+            fileData = fileData.replace("module top(", "module " + fileName + "(")
+        else:
+            fileData = fileData.replace("module " + fileName[0:len(fileName)-5] + "\n", "module " + fileName)
+        fin.close()
+        fin = open(PATH + fileName + ".v", "w")
```
Comment (jgoeders) [inline scripts/bfasst/compare/waveform.py:91]: Use `with` when opening files

File: scripts/bfasst/compare/waveform.py [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+        fin.write(fileData)
+        fin.close()
+
+    def parse(self, file):
+        netlist = sdn.parse(file)
+        library = netlist.libraries[0]
+        definition = library.definitions[0] #CHANGE THIS FOR MULTIPLE-MODULE DESIGNS.
+
+        i = 0
+        while(i < len(definition.ports)):
```
Comment (jgoeders) [inline scripts/bfasst/compare/waveform.py:101]: In python you would use:
`for port in definition.ports`

File: scripts/bfasst/compare/waveform.py [hunk 7]
```diff
@@ -0,0 +1,355 @@
...
+                data["outputList"].append(definition.ports[i].name)
+                data["totalList"].append(definition.ports[i].name)
+                data["oBitsList"].append(len(definition.ports[i].pins)-1)
+            elif (str(definition.ports[i].direction) == "Direction.IN"):
+                data["inputList"].append(definition.ports[i].name)
+                data["totalList"].append(definition.ports[i].name)
+                data["iBitsList"].append(len(definition.ports[i].pins)-1)
+            i = i + 1
+
+    def parseReversed(self, path, file):
```
Comment (jgoeders) [inline scripts/bfasst/compare/waveform.py:112]: Can you add docstrings to functions and some more comments to this file?

File: scripts/bfasst/compare/waveform.py [hunk 8]
```diff
@@ -0,0 +1,355 @@
...
+        file = open(file)
+        if(exists(path + "test.v")):
+            os.remove(path + "test.v")
```
Comment (jgoeders) [inline scripts/bfasst/compare/waveform.py:115]: Use pathlib

File: scripts/bfasst/compare/waveform.py [hunk 9]
```diff
@@ -0,0 +1,355 @@
...
+        newFile = open(path + "test.v", "x")
```
Comment (jgoeders) [inline scripts/bfasst/compare/waveform.py:116]: Use `with`

File: scripts/bfasst/compare/waveform.py [hunk 10]
```diff
@@ -0,0 +1,355 @@
...
+            elif "output" in line:
+                newFile.write(line)
+            else:
+                if (i == 0):
+                    i = 1
+                    newFile.write(line)
+        file.close()
+        newFile.close()
+        self.parse(newFile.name)
+        os.remove(path + "test.v")
```
Comment (jgoeders) [inline scripts/bfasst/compare/waveform.py:133]: use pathlib

File: scripts/bfasst/compare/waveform.py [hunk 11]
```diff
@@ -0,0 +1,355 @@
...
+            line = ("module " + fileName[0] + "_tb;")
+            tb.write(line)
+            line = "\n"
+        
+        if "TB_NAME)" in line:
+            line=("    $dumpvars(0," + fileName[0] + "_tb);\n")
+        
+        if "INPUTS" in line:
+            i = 0
+            while(i < self.inputNum()):
```
Comment (jgoeders) [inline scripts/bfasst/compare/waveform.py:146]: Don't use loops like this in python.  Either iterate through the object list, or if you need a numbered iterator use:
`for i in range(self.inputNum())`

File: scripts/bfasst/compare/waveform.py [hunk 12]
```diff
@@ -0,0 +1,355 @@
...
+                if data["inputList"][i] != "clk":
+                    line = "reg [" + str(data["iBitsList"][i]) + ":0] " + data["inputList"][i] + " = 0;\n"
+                    tb.write(line)
+                i = i+1
+            line = ""
+
+        if "OUTPUTS" in line:
+            i = 0
+            while(i < self.outputNum()):
```
Comment (jgoeders) [inline scripts/bfasst/compare/waveform.py:155]: same as above

File: scripts/bfasst/compare/waveform.py [hunk 13]
```diff
@@ -0,0 +1,355 @@
...
+                line = "wire [" + str(data["oBitsList"][i]) + ":0] " + data["outputList"][i] + ";\n"
+                tb.write(line)
+                i=i+1
+            line = ""
+        
+        if "MODULE_NAME" in line:
+            i = 0
+            line = fileName[0] + " instanceOf ("
+            while(i < self.totalNum()):
```
Comment (jgoeders) [inline scripts/bfasst/compare/waveform.py:164]: same as above

File: scripts/bfasst/compare/waveform.py [hunk 14]
```diff
@@ -0,0 +1,355 @@
...
+                if(i == self.totalNum() - 1):
+                    line = line + data["totalList"][i] + ");\n"
+                else:
+                    line = line + data["totalList"][i] + ", "
+                i = i + 1
+
+        if "/*SIGNALS" in line:
+            i = 0
+            while(i < self.inputNum()):
```
Comment (jgoeders) [inline scripts/bfasst/compare/waveform.py:173]: same as above

File: scripts/bfasst/compare/waveform.py [hunk 15]
```diff
@@ -0,0 +1,355 @@
...
+                if data["iBitsList"][i] == 0:
+                    data["randList"].append(np.random.randint(low = 0, high = 2, size = int(tests)))
+                else:
+                    data["randList"].append(np.random.randint(low = 0, high = (2**(int(data["iBitsList"][i]) + 1)-1), size = int(tests)))
+                i=i+1
+            i = 0
+            j = 0
+            while(i < int(tests)):
```
Comment (jgoeders) [inline scripts/bfasst/compare/waveform.py:181]: same as above

File: scripts/bfasst/compare/waveform.py [hunk 16]
```diff
@@ -0,0 +1,355 @@
...
+        if fileName[fileNum-1] + "_tb;" in line:
+            line = line.replace(fileName[fileNum-1], fileName[fileNum])
+        
+        if fileName[fileNum-1] + "_tb);" in line:
+            line = "    $dumpvars(0," + fileName[fileNum] + "_tb);\n"
+        
+        if fileName[fileNum-1] + " instanceOf (" in line:
+            line = fileName[fileNum] + " instanceOf ("
+            i = 0
+            while(i < self.totalNum()):
```
Comment (jgoeders) [inline scripts/bfasst/compare/waveform.py:206]: same as above

File: scripts/bfasst/compare/waveform.py [hunk 17]
```diff
@@ -0,0 +1,355 @@
...
+                if(i == self.totalNum() - 1):
+                    line = line + data["totalList"][i] + ");\n"
+                else:
+                    line = line + data["totalList"][i] + ", "
+                i = i + 1
+
+        tb.write(line)
+
+    def generateFirstTCL(self, PATH, fileName, fileNum):
```
Comment (jgoeders) [inline scripts/bfasst/compare/waveform.py:215]: don't use capitalized variable unless it's a global

File: scripts/bfasst/compare/waveform.py [hunk 18]
```diff
@@ -0,0 +1,355 @@
...
+        if(exists(PATH + fileName[fileNum] + ".tcl")):
+            os.remove(PATH+ fileName[fileNum] + ".tcl")
+        TCL = open(PATH + fileName[fileNum] + ".tcl", "x")
+        i=0
+        line = "set filter [list "
+        while(i < self.totalNum()):
```
Comment (jgoeders) [inline scripts/bfasst/compare/waveform.py:221]: same as above

File: scripts/bfasst/compare/waveform.py [hunk 19]
```diff
@@ -0,0 +1,355 @@
...
+            line = line + fileName[fileNum] + "_tb." + str(data["totalList"][i]).strip() + " "
+            i=i+1
+        line = line + "]\n"
+        TCL.write(line)
+        TCL.write("gtkwave::addSignalsFromList $filter\n")
+        TCL.write('gtkwave::/File/Export/Write_VCD_File_As "' + str(PATH) + fileName[fileNum] + '.vcd"\n')
+        TCL.write("gtkwave::File/Quit")
+
+    def generateTCL(self, PATH, fileName, fileNum):
```
Comment (jgoeders) [inline scripts/bfasst/compare/waveform.py:230]: don't capitalize

File: scripts/bfasst/compare/waveform.py [hunk 20]
```diff
@@ -0,0 +1,355 @@
...
+        if(exists(PATH + fileName[fileNum] + ".tcl")):
+            os.remove(PATH + fileName[fileNum] + ".tcl")   
+        TCL = open(PATH + fileName[fileNum] + ".tcl", "x")
```
Comment (jgoeders) [inline scripts/bfasst/compare/waveform.py:233]: use `with`

File: scripts/bfasst/compare/waveform.py [hunk 21]
```diff
@@ -0,0 +1,355 @@
...
+        
+
+    def generate_files(self, design):
+        TESTS = 100 #Note: change this number to however many tests the testbench should run.
+        SAMPLE_PATH = str(pathlib.Path.cwd()) + "/scripts/bfasst/compare_waveforms/sample_tb.v"
+        fileName = [design.impl_netlist_path.name[0: len(design.impl_netlist_path.name)-2], 
+        design.reversed_netlist_path.name[0: len(design.reversed_netlist_path.name)-2]]
+        file = [open(design.impl_netlist_path), open(design.reversed_netlist_path)]
+        fileNum = 0
+        PATH = str(design.impl_netlist_path)
```
Comment (jgoeders) [inline scripts/bfasst/compare/waveform.py:248]: Use pathlib.Path objects

File: scripts/bfasst/compare/waveform.py [hunk 22]
```diff
@@ -0,0 +1,355 @@
...
+
+            if(x.name.find("reversed") != -1):
+                self.fix_file(PATH, fileName[fileNum], True)
+                self.parseReversed(PATH, x.name)
+            else:
+                self.fix_file(PATH, fileName[fileNum], False)
+                self.parse(x.name)
+            
+            if(exists(PATH + fileName[fileNum] + "_tb.v")):
+                os.remove(PATH + fileName[fileNum] + "_tb.v")
```
Comment (jgoeders) [inline scripts/bfasst/compare/waveform.py:261]: use pathlib

File: scripts/bfasst/compare/waveform.py [hunk 23]
```diff
@@ -0,0 +1,355 @@
...
+            fileNum = fileNum + 1
+            refresh(data)
+
+    def runTest(self, design):
+        isEquivalent = False
+        IMPL_NAME = design.impl_netlist_path.name[0: len(design.impl_netlist_path.name)-2]
+        PATH = str(design.impl_netlist_path)
+        PATH = PATH[0: len(PATH) - len(design.impl_netlist_path.name)]
+        string = "iverilog -o " + PATH + "dsn " + PATH + IMPL_NAME + "_tb.v " 
+        string = string + PATH + design.impl_netlist_path.name + " " + str(pathlib.Path.cwd()) + "/third_party/yosys/techlibs/xilinx/cells_sim.v"
```
Comment (jgoeders) [inline scripts/bfasst/compare/waveform.py:292]: Do we have something in `paths` for this already?

File: scripts/bfasst/compare/waveform.py [hunk 24]
```diff
@@ -0,0 +1,355 @@
...
+        os.system(string)
```
Comment (jgoeders) [inline scripts/bfasst/compare/waveform.py:293]: Don't use os.system.  Use subprocess.  See other bfasst modules

File: scripts/bfasst/compare/waveform.py [hunk 25]
```diff
@@ -0,0 +1,355 @@
...
+        os.system("diff " + PATH + IMPL_NAME + ".vcd " + PATH + REVERSED_NAME + ".vcd >> " + PATH + "diff.txt")
+
+        file = open(PATH + "diff.txt")
+        lines = 0
+        for line in file:
+            lines = lines + 1
+
+        if(lines > 8):
+            print("NOT EQUIVALENT! See " + PATH + "diff.txt for more info!")
+            os.remove(PATH + IMPL_NAME + "_temp.vcd")
```
Comment (jgoeders) [inline scripts/bfasst/compare/waveform.py:316]: Dont use "PATH +"

Use the pathlib "/" operator

File: scripts/bfasst/compare_waveforms/generate_tb.py [hunk 26]
```diff
@@ -0,0 +1,258 @@
+import os
+from random import randint
+import sys
+import os
+from os.path import exists
```
Comment (jgoeders) [inline scripts/bfasst/compare_waveforms/generate_tb.py:5]: Dont use os

**Review by jgoeders — CHANGES_REQUESTED** [review]

File: scripts/bfasst/compare/waveform.py (hunk 23 above)
Comment (JakeEdvenson) [reply scripts/bfasst/compare/waveform.py:292]: I added to paths to have the build folder and the xilinx_yosys_waveform build folder inside of it. Now, I can call 
`bfasst.paths.WAVEFORM_BUILD / design.rel_path` to achieve the same result as `PATH = str(design.impl_netlist_path)
        PATH = PATH[0: len(PATH) - len(design.impl_netlist_path.name)]`

File: scripts/bfasst/compare/waveform.py [hunk 27]
```diff
@@ -0,0 +1,541 @@
...
+    are different, the designs must be unequivalent. Removes all unnecessary files after testing."""
+
+    def run_test(self, design):
+        is_equivalent = False
+        build_dir = bfasst.paths.WAVEFORM_BUILD / design.rel_path
+        impl_path = design.impl_netlist_path
+        reversed_path = design.reversed_netlist_path
+        impl_module = impl_path.name[0 : len(impl_path.name) - 2]
+        reversed_module = reversed_path.name[0 : len(reversed_path.name) - 2]
+        dsn = build_dir / Path("dsn")
```
Comment (jgoeders) [inline scripts/bfasst/compare/waveform.py:428]: You don't have to wrap the string in a Path object.

File: scripts/bfasst/compare/waveform.py [hunk 28]
```diff
@@ -0,0 +1,541 @@
...
+                            file_name[file_num - 1], file_name[file_num]
+                        )
+                    TCL.write(line)
+
+    """The main function that generates testbenches and TCL files. It begins by calling the parsers for the input & output names, then
+    it calls the testbench generators, finally it calls the TCL generators. It then increments to the next file and clears the data structure."""
+
+    def generate_files(self, design):
+        test_num = 100
+        sample_path = bfasst.paths.ROOT_PATH / Path(
```
Comment (jgoeders) [inline scripts/bfasst/compare/waveform.py:361]: No Path needed

File: scripts/bfasst/compare/waveform.py [hunk 29]
```diff
@@ -0,0 +1,541 @@
...
+        reversed_module = reversed_path.name[0 : len(reversed_path.name) - 2]
+        file_name = [impl_module, reversed_module]
+
+        # opens both files to be used in generating respective testbenches and TCLs
+        with open(impl_path) as impl_file:
+            with open(reversed_path) as reversed_file:
+                file = [impl_file, reversed_file]
+                build_dir = bfasst.paths.WAVEFORM_BUILD / design.rel_path
+                for f, file_num in zip(file, range(len(file))):
+                    file_path = build_dir / Path(file_name[file_num] + ".v")
```
Comment (jgoeders) [inline scripts/bfasst/compare/waveform.py:376]: Remove Path()

File: scripts/bfasst/compare/waveform.py [hunk 30]
```diff
@@ -0,0 +1,541 @@
...
+                        self.fix_file(file_path, file_name[file_num], False)
+                        self.parse(f.name)
+
+                    if tb_path.exists():
+                        tb_path.unlink()
+
+                    if file_num == 0:
+                        compare_path = sample_path
+                    else:
+                        compare_path = build_dir / Path(
```
Comment (jgoeders) [inline scripts/bfasst/compare/waveform.py:393]: Remove Path()

File: scripts/bfasst/compare_waveforms/generate_tb.py [hunk 31]
```diff
@@ -0,0 +1,258 @@
...
+    "outputList": [],
+    "oBitsList": [],
+    "totalList": [],
+    "randList": []
+}
+
+fileName = []
+file = []
+fileNum = 0
+PATH = "/home/edvenson/bfasst/build/waveform_comparison/"
```
Comment (jgoeders) [inline scripts/bfasst/compare_waveforms/generate_tb.py:22]: hardcoded paths?

File: scripts/bfasst/compare_waveforms/generate_tb.py [hunk 32]
```diff
@@ -0,0 +1,258 @@
...
+    TCL = open(PATH + fileName[fileNum] + ".tcl", "x")
+    sample = open(PATH + fileName[fileNum-1] + ".tcl")
+    for line in sample:
+        if fileName[fileNum-1] in line:
+            line = line.replace(fileName[fileNum-1], fileName[fileNum])
+        TCL.write(line)
+
+
+def main(fileList):
+    PATH = "/home/edvenson/bfasst/build/waveform_comparison/"
```
Comment (jgoeders) [inline scripts/bfasst/compare_waveforms/generate_tb.py:222]: hardcoded paths

File: scripts/bfasst/paths.py [hunk 33]
```diff
@@ -8,7 +8,10 @@
 RESOURCES_PATH = ROOT_PATH / "resources"
 SCRIPTS_PATH = ROOT_PATH / "scripts"
 ERROR_FLOW_PATH = ROOT_PATH / "error_flows"
+BUILD_PATH = ROOT_PATH / "build"
```
Comment (jgoeders) [inline scripts/bfasst/paths.py:11]: Use  self.work_dir instead of config.BUILD_PATH

**Review by jgoeders — CHANGES_REQUESTED** [review]

**Pull request merged**